### PR TITLE
numpy 1.12.0rc2 compatibility fix related to variable typecasting

### PR DESCRIPTION
**Pull request by chinnadhurai**
Fix for TypeError: Cannot cast array data from dtype('float32') to dtype('int64') according to the rule 'safe' can be fixed by making the cast explicit, casting="unsafe".

related to [5396](https://github.com/Theano/Theano/issues/5396)

### CODE REVIEW

File: theano/sparse/sandbox/sp2.py [hunk 1]
```diff
@@ -116,6 +117,11 @@ def make_node(self, n, p, shape):
         n = tensor.as_tensor_variable(n)
         p = tensor.as_tensor_variable(p)
         shape = tensor.as_tensor_variable(shape)
+
+        assert n.dtype in discrete_dtypes
+        assert p.dtype not in discrete_dtypes
```
Comment (slefrancois) [inline theano/sparse/sandbox/sp2.py]: p is a probability, shouldn't it be a float?

**Comment by slefrancois** [comment]
jenkins test this

File: theano/sparse/sandbox/sp2.py (hunk 1 above)
Comment (chinnadhurai) [inline theano/sparse/sandbox/sp2.py]: yeah..so, change "assert p.dtype not in discrete_dtypes" to "assert p.dtype in float_dtypes" ?

Comment (slefrancois) [inline theano/sparse/sandbox/sp2.py]: Let's try that, it will avoid user errors.

Comment (chinnadhurai) [inline theano/sparse/sandbox/sp2.py]: sure, committed the changes.. also confirmed that the nosetests still pass

**Comment by slefrancois** [comment]
jenkins test this

**Comment by slefrancois** [comment]
@nouiz , looks good to me once jenkins passes, can you double-check?

**Comment by nouiz** [comment]
ok for me too.

**Pull request merged**